File: npm-bazel-gen.py
import json, subprocess, os, re, shutil, sys
from Queue import Queue
from threading import Thread

registry = "https://registry.npmjs.org/"

packages = subprocess.check_output(['git', 'ls-files', '*/package.json']).strip().split('\n')
#print '\n'.join(packages)

internal_modules = {}
external_dependencies = set()

print("-- scanning packages")
for package_file in packages:
	print(package_file)
	package_dir = os.path.dirname(package_file)
	with open(package_file) as x: package = json.load(x)
	internal_modules[package["name"]] = package_dir
	if "dependencies" in package:
		for name, version in package["dependencies"].viewitems():
			external_dependencies.add(name + "@" + version)
	if "devDependencies" in package:
		for name, version in package["devDependencies"].viewitems():
			external_dependencies.add(name + "@" + version)

version_cache = {}
dependency_cache = {}

if os.path.exists('build_tools/npm_version_cache'):
	with open('build_tools/npm_version_cache') as x: version_cache = json.load(x)
if os.path.exists('build_tools/npm_dependency_cache'):
	with open('build_tools/npm_dependency_cache') as x: dependency_cache = json.load(x)

workspace_urls = {}

def is_url(str):
	return str.startswith('http://') or str.startswith('https://')

def get_version(name, version):
	code = name + '@' + version
	if re.match(r'^[\d.]+\.[\d.]+\.[\d.]+$', version):
		#print("  -- version seems exact: " + code)
		return version
	elif is_url(version):
		#print("  -- version seems to be a url: " + code)
		return version
	elif code in version_cache:
		#print("  -- version cache hit: " + code + " = " + version_cache[code])
		return version_cache[code]
	print("  -- version: " + code)
	output = subprocess.check_output(['npm', '--registry', registry, 'view', code, 'version']).strip().split('\n')
	if len(output) > 1:
		last_line = output[-1]
		new_code = last_line.split(' ')[0]
		resolved_version = re.match(r'^(.+)@([^@]+)$', new_code).groups()[1]
	else:
		resolved_version = output[0]
	print("    -- recording {}@{} = {}".format(name, version, resolved_version))
	version_cache[code] = resolved_version
	with open('build_tools/npm_version_cache', 'w') as x: json.dump(version_cache, x, indent=2, separators=(',', ': '), sort_keys=True)
	return resolved_version

def get_dependencies(name, version):
	code = name + '@' + version
	if code in dependency_cache:
		return dependency_cache[code]
	print("  -- dependencies: " + code)
	result_string = subprocess.check_output(['npm', '--registry', registry, 'view', '--json', name+'@'+version, 'dependencies'])
	if not result_string:
		output = {}
	else:
		try:
			output = json.loads(result_string)
		except ValueError as e:
			raise Exception("error reading dependencies for {}@{}: <{}>\nCaused by: {}: {}".format(name, version, result_string, type(e).__name__, str(e))), None, sys.exc_info()[2]

	print("    -- " + json.dumps(output))
	dependency_cache[code] = output
	with open('build_tools/npm_dependency_cache', 'w') as x: json.dump(dependency_cache, x, indent=2, separators=(',', ': '), sort_keys=True)
	return output

def get_rule_name(name, version):
	raw_key = name + "_" + version;
	if is_url(version):
		raw_key = name + "_tarball"
	return re.sub(r'[^A-Za-z0-9_]+', "_", raw_key)

def add_workspace_url(name, version):
	rule_name = get_rule_name(name, version)
	if is_url(version):
		workspace_urls[rule_name] = version
	else:
		workspace_urls[rule_name] = '{repo_url}/{name}/-/{name}-{version}.tgz'.format(
			repo_url = registry,
			name = name,
			version = version
		)

def get_transitive_dependencies(parent_name, parent_version, dependencies):
	add_workspace_url(parent_name, parent_version)
	for name, version in get_dependencies(parent_name, parent_version).viewitems():
		resolved_version = get_version(name, version)
		add_workspace_url(name, resolved_version)
		code = name+'@'+resolved_version
		if code in dependencies:
			continue
		dependencies.add(code)
		get_transitive_dependencies(name, resolved_version, dependencies)

print('-- writing thirdparty BUILD file')

external_npm_module_template = """external_npm_module(
	name='{}@{}',
	raw_target='{}',{}
	visibility = ["//visibility:public"],
)
"""

def worker():
	while True:
		parent_name, parent_version, dependencies = q.get()
		resolved_parent_version = get_version(parent_name, parent_version)
		for name, version in get_dependencies(parent_name, resolved_parent_version).viewitems():
			resolved_version = get_version(name, version)
			code = name+'@'+resolved_version
			if code in dependencies:
				continue
			dependencies.add(code)
			q.put((name, resolved_version, dependencies))
		q.task_done()

q = Queue()

num_worker_threads = 16

for i in range(num_worker_threads):
	t = Thread(target=worker)
	t.daemon = True
	t.start()

resolved_external_dependencies = set()
for code in external_dependencies:
	name, version = re.match(r'^(.+)@([^@]+)$', code).groups()
	if name in internal_modules:
		continue
	if is_url(version):
		version = "tarball"
	resolved_version = get_version(name, version)
	resolved_external_dependencies.add(name+'@'+resolved_version)
	transitive_dependencies = set()
	q.put((name, version, transitive_dependencies))

q.join()
if not os.path.exists('build_tools/npm-thirdparty'):
	os.makedirs('build_tools/npm-thirdparty')

with open('build_tools/npm-thirdparty/BUILD', 'w') as BUILD:
	BUILD.write("load('/build_tools/npm', 'external_npm_module')\n\n");
	for code in sorted(resolved_external_dependencies):
		name, version = re.match(r'^(.+)@([^@]+)$', code).groups()
		print("-- " + code)
		transitive_dependencies = set()
		get_transitive_dependencies(name, version, transitive_dependencies)
		raw_target = "@{}//:raw".format(get_rule_name(name, version))
		depstring = ""
		if len(transitive_dependencies):
			deps = []
			for dep in sorted(transitive_dependencies):
				dep_name, dep_version = re.match(r'^(.+)@([^@]+)$', dep).groups()
				deps.append("'@{}//:raw'".format(get_rule_name(dep_name, dep_version)))
			depstring = "\n\truntime_deps=[\n\t\t" + ",\n\t\t".join(deps) + "\n\t],\n"
		BUILD.write(external_npm_module_template.format(
			name,
			version,
			raw_target,
			depstring
		))

print('-- writing workspace')

workspace_template = """new_http_archive(
	name='{rule_name}',
	url='{url}',
	build_file_content="filegroup(name='raw', srcs=glob(['*'], exclude_directories=0), visibility=['//visibility:public'])",
)
"""

if sys.platform == 'darwin':
	node_platform = 'darwin'
	phantom_platform = 'macosx'
	phantom_file_extension = 'zip'
else:
	node_platform = 'linux'
	phantom_platform = 'linux-x86_64'
	phantom_file_extension = 'tgz'

workspace_preamble = """
new_http_archive(
	name='node',
	url='https://nodejs.org/download/release/v4.3.1/node-v4.3.1-{node_platform}-x64.tar.gz',
	strip_prefix='node-v4.3.1-{node_platform}-x64',
	build_file_content="filegroup(name='raw', data=glob(['*'], exclude_directories = 0), visibility=['//visibility:public'])"
)

new_http_archive(
	name='phantomjs',
	url='http://thirdpartyrepository.redfintest.com/com/redfin/phantomjs/2.1.1/phantomjs-2.1.1-{phantom_platform}.{phantom_file_extension}',
	strip_prefix='phantomjs-2.1.1-{phantom_platform}',
	build_file_content="filegroup(name='executable', data=glob(['bin/phantomjs']), visibility=['//visibility:public'])"
)

new_http_archive(
	name='node_headers',
	url='https://nodejs.org/download/release/v4.3.1/node-v4.3.1-headers.tar.gz',
	build_file_content=r\"""
genrule(
	name='gyp-package',
	srcs=glob(['**']),
	outs=['gyp-package.tar.gz'],
	visibility=['//visibility:public'],
	cmd='''
rm -rf $(@D)/.node-gyp
mkdir -p $(@D)/.node-gyp
cp -r external/node_headers/node-v4.3.1 $(@D)/.node-gyp/4.3.1
echo 9 > $(@D)/.node-gyp/4.3.1/installVersion
tar zcf $@ -C $(@D) .node-gyp
'''
)
\"""
)


""".format(node_platform=node_platform, phantom_platform=phantom_platform, phantom_file_extension=phantom_file_extension)



with open('WORKSPACE', 'w') as WORKSPACE:
	WORKSPACE.write(workspace_preamble)
	for rule_name in sorted(workspace_urls.keys()):
		url = workspace_urls[rule_name]
		WORKSPACE.write(workspace_template.format(rule_name=rule_name, url=url))

print("-- writing internal BUILD files")
for package_file in packages:
	package_dir = os.path.dirname(package_file)
	with open(package_file) as x: package = json.load(x)
	dependencies = []
	if "dependencies" in package:
		for name in sorted(package["dependencies"].keys()):
			dependencies.append((name, package["dependencies"][name]))
	if "devDependencies" in package:
		for name in sorted(package["devDependencies"].keys()):
			dependencies.append((name, package["devDependencies"][name]))

	with open(package_dir+'/BUILD', 'w') as BUILD:
		BUILD.write("load('/build_tools/npm', 'internal_npm_module')\n\n");
		BUILD.write("internal_npm_module(\n    name='")
		currentName = os.path.basename(package_dir)
		BUILD.write(currentName)
		BUILD.write("',\n    srcs=glob(['**'], exclude=['node_modules/**', 'target/**'")
		# add custom exclusions here
		BUILD.write("]),\n    package_json='"+os.path.basename(package_file)+"',\n    ")
		# configure custom install/pack scripts here
		BUILD.write("deps=[\n")
		if "dependencies" in package:
			for name in sorted(package["dependencies"].keys()):
				version = package["dependencies"][name]
				if name in internal_modules:
					BUILD.write("        '//{}',\n".format(internal_modules[name]))
				else:
					resolved_version = get_version(name, version)
					BUILD.write("        '//build_tools/npm-thirdparty:{}@{}',\n".format(name, resolved_version))
		BUILD.write("    ],\n    dev_deps = [\n")
		if "devDependencies" in package:
			for name in sorted(package["devDependencies"].keys()):
				version = package["devDependencies"][name]
				if name in internal_modules:
					BUILD.write("        '//{}',\n".format(internal_modules[name]))
				else:
					resolved_version = get_version(name, version)
					BUILD.write("        '//build_tools/npm-thirdparty:{}@{}',\n".format(name, resolved_version))
		BUILD.write("    ],\n    visibility=['//visibility:public'],")
		BUILD.write("\n)\n")

		# add extra custom rules here

File: npm-bazel-gen.js
"use strict";

var semver = require('semver');
var path = require('path');
var child_process = require('child_process');
var http = require('http');
var fs = require('fs');
var crypto = require('crypto');
var dns = require('dns');

// limit the number of sockets npm-bazel gen opens to try
// to help avoid overloading sinopia.
var agent = new http.Agent({maxSockets: 40});

var debugLogging = true;
mkdirp('tools/cache');
var packageHashFile = 'tools/cache/npm_package_hashes';
var registryCacheFile = 'tools/cache/npm_registry';

var packages = child_process.execFileSync('git', ['ls-files', '*/package.json'], {encoding: 'utf8'}).trim().split('\n');

var internalModules = {};
var externalDependencies = {};

var nameVersionPairs = {};
function NameVersionPair(name, version) {
	this.name = name;
	this.version = version;
	this.hashCode = name + "@" + version;
	nameVersionPairs[this.hashCode] = this;
}

var packageHashes = {};
const nbg_js = path.relative('.', __filename);

packages.forEach((packageFile) => {
	if (debugLogging) console.log(packageFile);
	var packageBody = fs.readFileSync(packageFile, 'utf8');
	var packageDir = path.dirname(packageFile);
	var packageDescriptor = JSON.parse(packageBody);
	internalModules[packageDescriptor.name] = packageDir;

	var deps = packageDescriptor.dependencies || {};
	Object.assign(deps, packageDescriptor.devDependencies || {});
	Object.keys(deps).forEach((name) => {
		var version = deps[name];
		var pair = new NameVersionPair(name, version);
		externalDependencies[pair.hashCode] = true;
	});
});

var registryCache = {};
var inFlight = {};
var fetchedAnything = false;
var crawled = {};

var workspaceUrls = {};
var versionCache = {};

var transitiveDependencyCache = {};

var npmIp = null;
var npmDnsInflight = [];

console.log("-- fetching from registry");

try {
	registryCache = JSON.parse(fs.readFileSync(registryCacheFile, 'utf8'));
} catch (e) {
	if (e.code !== "ENOENT") throw e;
}

inFlight.LOADING = () => {};
Object.keys(externalDependencies).forEach(function(dep) {
	var pair = nameVersionPairs[dep];
	if (internalModules[pair.name]) return;
	fetchFromRegistry(pair.name, () => crawlPackageJson(pair.name, pair.version));
});
delete inFlight.LOADING;
(() => {
	var inFlightCount = Object.keys(inFlight).length;
	if (debugLogging) console.log("in flight: " + inFlightCount);
	if (!inFlightCount) {
		doneCrawling();
	}
})();

function die(msg) {
	console.error(msg);
	process.exit(1);
}

function dnsLookup(callback) {
	if (npmIp) {
		return callback(npmIp);
	}
	npmDnsInflight.push(callback);
	if (npmDnsInflight.length > 1) {
		return;
	}
	dns.lookup('registry.npmjs.org', (err, result) => {
		if (err) {
			console.error(err);
			throw err;
		}
		console.log("dns success");
		npmIp = result;
		for (callback of npmDnsInflight) callback(npmIp);
	});
}

function fetchFromRegistry(name, callback) {
	if (inFlight[name]) {
		return inFlight[name].push(callback);
	}
	inFlight[name] = [callback];
	if (registryCache[name]) {
		return resolve();
	}
	console.log("  fetching " + name);
	dnsLookup((ip) => {
		var uripath = '/' + encodeURIComponent(name);
		http.request({agent: agent, host:ip, headers: {"Host":"registry.npmjs.org"}, path:uripath}, (res) => {
			if (res.statusCode != 200) die(`request for ${path} failed with status code ${res.statusCode}`);
			var body = "";
			res.on('data', function (chunk) {
				body += chunk;
			});

			res.on('end', function () {
				registryCache[name] = JSON.parse(body);
				console.log("  fetched " + name);
				fetchedAnything = true;
				resolve();
			});
		}).end();
	});

	function resolve() {
		var observers = inFlight[name];
		delete inFlight[name];
		if (observers) observers.forEach((callback) => callback());
	}
}

function getPackageJsonFromCache(name, version) {
	var response = registryCache[name];
	var resolvedVersion = semver.maxSatisfying(Object.keys(response.versions), version);
	return response.versions[resolvedVersion];
}

function crawlPackageJson(name, version) {
	var packageJson;
	if (isUrl(version)) {
		packageJson = {};
	} else {
		packageJson = getPackageJsonFromCache(name, version);
		if (!packageJson) {
			throw new Error(`should have been in cache: ${name}@${version}`);
		}
	}
	var code = packageJson.name + "@" + packageJson.version;
	if (!crawled[code]) {
		if (debugLogging) console.log("crawling " + code);
		crawled[code] = true;
		if (packageJson.dependencies) {
			inFlight[code + " LOADING"] = () => {};
			Object.keys(packageJson.dependencies).forEach((name) => {
				if (internalModules[name]) return;
				var version = packageJson.dependencies[name];
				fetchFromRegistry(name, () => crawlPackageJson(name, version));
			});
			delete inFlight[code + " LOADING"];
		}
	}
	var inFlightCount = Object.keys(inFlight).length;
	if (debugLogging) console.log("in flight: " + inFlightCount);
	if (!inFlightCount) {
		doneCrawling();
	}
}

function getResolvedVersion(name, version) {
	if (isUrl(version)) return version;
	var response = registryCache[name];
	if (!response) {
		throw new Error(`Missing response in registry cache: ${name}@${version}`);
	}
	var resolvedVersion = semver.maxSatisfying(Object.keys(response.versions), version);
	return resolvedVersion;
}

function writeFileOrDie(fileName, text) {
	try {
		fs.writeFileSync(fileName, text, 'utf8');
	} catch (e) {
		die("Couldn't write " + fileName + ": " + e);
	}
}

function doneCrawling() {
	if (fetchedAnything) {
		console.log("-- writing registry cache");
		writeFileOrDie(registryCacheFile, JSON.stringify(registryCache, null, 2));
	}
	console.log("-- writing thirdparty BUILD file");
	writeThirdPartyBuildFile();
	console.log("-- writing WORKSPACE");
	writeWorkspace();
	console.log("-- writing internal BUILD files");
	writeInternalBuildFiles();
	console.log("-- writing package hashes");
	writeFileOrDie(packageHashFile, JSON.stringify(packageHashes));
}

function writeThirdPartyBuildFile() {
	var resolvedExternalDependencies = {};
	Object.keys(externalDependencies).sort().forEach((code) => {
		var pair = nameVersionPairs[code];
		if (internalModules[pair.name]) return;
		var resolvedVersion = getResolvedVersion(pair.name, pair.version);
		var resolvedPair = new NameVersionPair(pair.name, resolvedVersion);
		resolvedExternalDependencies[resolvedPair.hashCode] = true;
		versionCache[pair.hashCode] = resolvedVersion;
	});

	var buffer = ["load('/tools/npm', 'external_npm_module')\n\n"];
	Object.keys(resolvedExternalDependencies).sort().forEach((code) => {
		var pair = nameVersionPairs[code];
		buffer.push(`external_npm_module(
	name='${pair.hashCode}',
	tarball='@${getRuleName(pair.name, pair.version)}//file',`);
		var transitiveDependencies = getTransitiveDependencies(pair.name, pair.version, 0);
		if (Object.keys(transitiveDependencies).length) {
			buffer.push("\n\truntime_deps=[\n\t\t");
			var first = true;
			Object.keys(transitiveDependencies).sort().forEach((depCode) => {
				var depPair = nameVersionPairs[depCode];
				if (first) {
					first = false;
				} else {
					buffer.push(",\n\t\t");
				}
				buffer.push(`'@${getRuleName(depPair.name, depPair.version)}//file'`);
			})
			buffer.push("\n\t],\n");
		}
		buffer.push("\n\tvisibility = ['//visibility:public'],\n)\n\n");
	});
	mkdirp('tools/npm-thirdparty');
	writeFileOrDie('tools/npm-thirdparty/BUILD', buffer.join(''));
}

function isUrl(version) {
	return /^https?:\/\//.test(version);
}

function getRuleName(name, version) {
	var rawKey;
	if (/^https?:\/\//.test(version)) {
		rawKey = name + "_tarball";
	} else {
		rawKey = name + "_" + version;
	}
	return rawKey.replace(/[^A-Za-z0-9_]+/g, "_").replace(/^_*/, '');
}

function addWorkspaceUrl(name, version) {
	var ruleName = getRuleName(name, version);
	if (isUrl(version)) {
		workspaceUrls[ruleName] = version;
	} else {
		var uripath = encodeURIComponent(name);
		var nameonly = name.replace(/^.*\//, '');
		workspaceUrls[ruleName] = `https://registry.npmjs.org/${uripath}/-/${nameonly}-${version}.tgz`;
	}
}

function getTransitiveDependencies(name, version, depth) {
	var pair = new NameVersionPair(name, version);
	if (transitiveDependencyCache[pair.hashCode]) {
		return transitiveDependencyCache[pair.hashCode];
	}
	if (debugLogging) console.log("  ".repeat(depth) + `${name}@${version}`);
	addWorkspaceUrl(name, version);
	var packageJson = getPackageJsonFromCache(name, version);
	var dependencies = {};
	if (packageJson && packageJson.dependencies) {
		dependencies = packageJson.dependencies;
	}
	var resultSet = {};
	transitiveDependencyCache[pair.hashCode] = resultSet;
	Object.keys(dependencies).forEach((childName) => {
		if (internalModules[childName]) return;
		var child = new NameVersionPair(childName, dependencies[childName]);
		try {
			var resolvedVersion = getResolvedVersion(childName, child.version);
		} catch (e) {
			console.error(`Error resolving dependency of ${name}@${version}`, e);
			throw e;
		}
		var resolvedChild = new NameVersionPair(childName, resolvedVersion);
		versionCache[child.hashCode] = resolvedVersion;
		addWorkspaceUrl(childName, resolvedVersion);
		resultSet[resolvedChild.hashCode] = true;
		var childResultSet = getTransitiveDependencies(childName, resolvedVersion, depth+1);
		Object.keys(childResultSet).forEach((key) => {
			resultSet[key] = true;
		});
	});
	return resultSet;
}

function writeWorkspace() {
	var buffer = [`
workspace(name='redfin_main')

load('/tools/node_repository_rules', 'node_binary', 'node_headers')
node_binary(
	version = '6.6.0',
	shas = {
		'linux': 'c22ab0dfa9d0b8d9de02ef7c0d860298a5d1bf6cae7413fb18b99e8a3d25648a',
		'darwin': 'c8d1fe38eb794ca46aacf6c8e90676eec7a8aeec83b4b09f57ce503509e7a19f',
	},
)

node_binary(
	version = '4.3.1',
	shas = {
		'linux': 'b3af1ed18a9150af42754e9a0385ecc4b4e9b493fcf32bf6ca0d7239d636254b',
		'darwin': '9c0751ee88a47c10269eb930d7ad7b103c2ba875c3a96204ca133dc52fc50826',
	},
)

# //tools/node_version_select is actually an alias that resolves to different
# node versions depending on whether or not '--define=node=6' is passed to bazel.
bind(
	name='node',
	actual='//tools:node_alias',
)

node_headers(
	version = '6.6.0',
	sha = '60b81c7276105a51e71ad8bc7f59163105e7c5dd1d992b173b5b66449b6df3fc'
)

node_headers(
	version = '4.3.1',
	sha = '8ba5c1e5eb5509e0f4f00d56e1916ac703fdd05cf353f119451f2b37c51987a5',
)

bind(
	name = 'node_headers',
	actual = '//tools:node_headers_alias',
)

`];
	Object.keys(workspaceUrls).sort().forEach((ruleName) => {
		var url = workspaceUrls[ruleName];
		buffer.push(`http_file(
	name='${ruleName}',
	url='${url}',
)
`);
	});
	writeFileOrSkip('WORKSPACE', buffer.join(''));
}

function mkdirp(dir) {
	try {
		fs.mkdirSync(dir);
	} catch (e) {
		if (e.code === "ENOENT") {
			mkdirp(path.dirname(dir));
			fs.mkdirSync(dir);
		} else if (e.code !== "EEXIST") {
			throw e;
		}
	}
}

function writeInternalBuildFiles() {
	packages.forEach((packageFile) => {
		var packageDir = path.dirname(packageFile);
		var currentName = path.basename(packageDir);
		var packageDescriptor = JSON.parse(fs.readFileSync(packageFile, 'utf8'));

		var output;

		var result = getBUILDFileContent(currentName, packageFile, packageDescriptor);
		var buffer = result.buffer;
		var loadCommands = result.loadCommands;
		output = loadCommands.join("\n") + "\n\n" + buffer.join("");

		var outFile = packageDir + "/BUILD"

		// XXX IMPORTANT: this call is actually async, so don't do anything
		// in this method after it.
		writeFileOrSkip(outFile, output);
	});
}

function writeFileOrSkip(filePath, output) {
	fs.readFile(filePath, 'utf8', (err, body) => {
		// Bazel invalidates some of its caches when timestamps change
		if (body === output) return;
		fs.writeFile(filePath, output, 'utf8', (err) => {
			if (err) die("Couldn't write " + filePath + ": " + err);
		});
	})
}

function getBUILDFileContent(currentName, packageFile, packageDescriptor) {
	var TAB = "\t";
	var NEWLINE = "\n";

	var loadCommands = ["load('/tools/npm', 'internal_npm_module')"];
	var buffer = [`
internal_npm_module(
	name='${currentName}',
	`
	];
	buffer.push("srcs=glob(['**'], exclude=['BUILD', 'node_modules/**']),\n");
	buffer.push("\tpackage_json='"+path.basename(packageFile)+"',\n");
	buffer.push("\tdeps=[\n")

	buffer = buffer.concat(getDependenciesArray(packageDescriptor.dependencies));

	buffer.push("\t],\n\tdev_deps = [\n")

	buffer = buffer.concat(getDependenciesArray(packageDescriptor.devDependencies));

	buffer.push("\t],\n");
	buffer.push("\tvisibility=['//visibility:public'],\n)\n");

	return { loadCommands, buffer };

}

function getDependenciesArray(depsObj) {
	var TAB = "\t";
	var NEWLINE = "\n";
	var buffer = [];
	if (depsObj) {
		Object.keys(depsObj || {}).sort().forEach((name) => {
			var version = depsObj[name];
			if (internalModules[name]) {
				buffer.push(`${TAB}${TAB}'//${internalModules[name]}',${NEWLINE}`);
			} else {
				var resolvedVersion = getResolvedVersion(name, version);
				buffer.push(`${TAB}${TAB}'//tools/npm-thirdparty:${name}@${resolvedVersion}',${NEWLINE}`);
			}
		});
	}
	return buffer;
}

function safeReadFileSync(filename) {
	try {
		return fs.readFileSync(filename, 'utf8');
	} catch (e) {
		if (e.code !== "ENOENT") throw e;
	}
}

